patterns/strategy.md

```diff
@@ -0,0 +1,161 @@
+# Strategy (aka Policy)
+
+## Description
+
+The [Strategy design pattern](https://en.wikipedia.org/wiki/Strategy_pattern) is a technique that enables separation of concerns in object-oriented software design.
+It also allows to decouple software modules through [Dependency Inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle).
+The basic idea behind the Strategy pattern is that, given an algorithm solving a particular problem, we define only the skeleton of the algorithm at an abstract level and we delegate the specific algorithm’s implementation (or some of its parts) to another class or method.
+
+In this way, a client using the algorithm may choose a specific implementation, while the general algorithm workflow remains the same.
+In other words, the abstract specification of the class does not depend on the specific implementation of the derived class, but specific implementation must adhere to the abstract specification.
+This is why we call it "Dependency Inversion".
+
+## Motivation
+
+Imagine we are working on a project that generates reports every month.
+We need the reports to be generated in different formats (strategies), e.g.,
+in `JSON` or `Plain Text` formats.
+But things vary over time and we don't know what kind of requirement we may get in the future.
+For example, we may need to generate our report in a completly new format,
+or just modify one of the existing formats.
+
+## Example
+
+In this example our invariants (or abstractions) are `Context`, `Formatter`, and `Report`,
+while `Text` and `Json` are our strategy structs.
+These strategies have to implement the `Formatter` trait.
+
+```rust
+use std::collections::HashMap;
+use std::fmt::Write as FmtWrite;
+use std::{error, result};
+type Result = result::Result<String, Box<dyn error::Error>>;
```

Comment:
I think that this is confusing. I would prefer to use `Result<String, Box<dyn error::Error>>` everywhere it is necessary, instead of creating a type alias.
What other reviewers think?

Reply:
Maybe we can call it something like `AnyResult` and we can remove the `use std::{error, result};` line?

Reply:
I added `use std::{error, result};` to avoid compiler errors because I used `Result`. Now that your offer `AnyResult` we can remove `use std::{error, result};`.

Reply:
Other reviewers hasn't sent yet anything concerning the alias for `Result<String, Box<dyn error::Error>>`. If it is really confusing I can remove and use `Result<String, Box<dyn error::Error>>`.

Reply:
yes, sorry, only `Result` is in the prelude apparently